Docs for daemons (#346)
rucio/documentation: docs/started/daemons.md

---
id: daemons
title: Rucio Daemons
sidebar_label: daemons
---

Rucio relies on several daemons (processes) to perform different logic.
Most of the daemons connect to the DB to read some data, perform some computation, 
and then write some data back into the DB.

Usually one daemon will create some work for another daemon and viceversa.
In Rucio realm, daemons communicate to others by the DB.

The following table represents a high level view of the responsibility of each of the daemons.

## Daemons

> Name | Domain|  Purpose |  Details

Reviewer: I think this will be the last nit-pick. Is it is possible to set this up with the a relative path? Using this method will redirect the user to a different site if they are looking at the gh docs mirror.

rucio-auditor | Replica | Find inconsistencies on storage, for example, dark data discovery | [Details](https://rucio.cern.ch/documentation/bin/rucio-auditor)
rucio-automatix | Replica | Used for testing: injects random data in RSEs to check liveness | [Details](https://rucio.cern.ch/documentation/bin/rucio-automatix)
rucio-bb8 | Replica | Rebalance data across RSEs | [Details](https://rucio.cern.ch/documentation/bin/rucio-bb8)
rucio-cache-client | Replica | Populates information of replicas on volatile storage  | [Details](https://rucio.cern.ch/documentation/bin/rucio-cache-client)
rucio-cache-consumer | Replica | Adds and deletes cache replicas to the Rucio catalog | [Details](https://rucio.cern.ch/documentation/bin/rucio-cache-consumer)
rucio-conveyor-finisher | Transfer | Updates Rucio internal state after the file transfer has finished | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-finisher)
rucio-conveyor-poller | Transfer | Polls updates from the transfer tool to check the transfer state | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-poller)

> rucio-conveyor-preparer | Transer | Prepares data transfers | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-preparer)

Reviewer 2: ```suggestion
rucio-conveyor-preparer | Transfer | Prepares data transfers | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-preparer)
```

rucio-conveyor-stager | Transfer | Issues staging (bring online) requests to tape RSEs  | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-stager)
rucio-conveyor-submitter | Transfer | Submit transfer requests to the transfer tool (prepares the transfer as well if the conveyor-preparer is not enabled) | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-submitter)
rucio-conveyor-throttler | Transfer | Queues transfer requests inside Rucio, applying limits, ex: only one transfer at a time, etc ... | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-throttler)
rucio-dark-reaper | Deletion | Deletes quarantined replicas | [Details](https://rucio.cern.ch/documentation/bin/rucio-dark-reaper)
rucio-dumper | Consistency | Dumps file lists. The rucio-auditor consumes these dumps to discover dark data | [Details](https://rucio.cern.ch/documentation/bin/rucio-dumper)
rucio-follower | Telemetry | Aggregates events affecting DIDs | [Details](https://rucio.cern.ch/documentation/bin/rucio-follower)

> rucio-hermes | Telemtry | Sends Rucio messages to external services (influxDB, OpenSearch, ActiveMQ, ...) | [Details](https://rucio.cern.ch/documentation/bin/rucio-hermes)

Reviewer 2: ```suggestion
rucio-hermes | Telemetry | Sends Rucio messages to external services (InfluxDB, OpenSearch, ActiveMQ, ...) | [Details](https://rucio.cern.ch/documentation/bin/rucio-hermes)
```

rucio-judge-cleaner | Rule | Cleans expired replication rules | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-cleaner)
rucio-judge-evaluator | Rule | Creates and evaluates replication rules based on their state (OK/REPL/STUCK)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-evaluator)
rucio-judge-injector | Rule | Asynchronously injects replication rules  | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-injector)
rucio-judge-repairer | Rule | Repairs stuck replication rules (STATE=STUCK) | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-repairer)
rucio-kronos | Telemetry | Consumes Rucio tracing messages, updates access time of replicas and access count of DIDs| [Details](https://rucio.cern.ch/documentation/bin/rucio-kronos)
rucio-minos | Replica | Reads list of physical file names (PFNs) declared bad and classifies them in: temporary unavailable  and permanently unavailable (to be recovered by the necromancer daemon)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-minos)

> rucio-minos-temporary-expiration | Replica | Moves back replicas TEMPORARY_UNAVAILABLE into AVAILABE state  | [Details](https://rucio.cern.ch/documentation/bin/rucio-minos-temporary-expiration)

Reviewer 2: ```suggestion
rucio-minos-temporary-expiration | Replica | Moves back replicas TEMPORARY_UNAVAILABLE into AVAILABLE state  | [Details](https://rucio.cern.ch/documentation/bin/rucio-minos-temporary-expiration)
```

rucio-necromancer | Deletion | Works on permantenly unavailable replicas, it tries to recover the data from other valid replicas if any, else declares the replica as lost  | [Details](https://rucio.cern.ch/documentation/bin/rucio-necromancer)

> rucio-oauth-manager | Auth/Authz | deletes expired access token and and oath sessions and refresh access tokens with refresh tokens  | [Details](https://rucio.cern.ch/documentation/bin/rucio-oauth-manager)

Reviewer 2: ```suggestion
rucio-oauth-manager | Auth/Authz | deletes expired access token and and OAuth sessions and refresh access tokens with refresh tokens  | [Details](https://rucio.cern.ch/documentation/bin/rucio-oauth-manager)
```

rucio-reaper | Deletion | Deletes replicas that dont' have locks anymore, i.e they have a tombstone set | [Details](https://rucio.cern.ch/documentation/bin/rucio-reaper)

> rucio-suspicious-replica-recoverer | Replica | Declares suspicios replicas as bad if they are found available on other RSEs, so necromancer will work on them | [Details](https://rucio.cern.ch/documentation/bin/rucio-replica-recoverer)

Reviewer 2: ```suggestion
rucio-suspicious-replica-recoverer | Replica | Declares suspicious replicas as bad if they are found available on other RSEs, so necromancer will work on them | [Details](https://rucio.cern.ch/documentation/bin/rucio-replica-recoverer)
```

> rucio-rse-decommissioner | Deletion | decomissions an RSE. The actions to perform are specified in decomissioning profiles (delete all data, move replicas, etc ...)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-rse-decomissioner)

Reviewer 2: ```suggestion
rucio-rse-decommissioner | Deletion | decommissions an RSE. The actions to perform are specified in decommissioning profiles (delete all data, move replicas, etc ...)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-rse-decommissioner)
```

### What happens when new data is added to an existing dataset that already has replicas?
The judge evaluator will keep track of new data added to datasets that are already replicated to trigger the necessary transfer requests to ensure all new data is copied to the RSEs.

### What is the purpose of the minos daemons?
An human operator can declare some datasets as temporarily unavailable due to maintenance, outages, etc ...The operator will set an expiration time on the temporary unavailable status. When the expiration time is reached, the minos-temporary-expiration will put the replicas back in available state.

### What is the relationship between auditor, rucio-dumper and dark-reaper?
The dumper will create a dump of all the files in an RSE that will be passed to the auditor. The auditor will check for inconsistencies and mark missing data as dark data (quarantined replicas).
Dark reaper is the one deleting this dark data to free up space from the quarantined replicas table.

### How is data deleted?
When replicas are healthy, the judge-cleaner will set a tombstone on replicas where the lifetime has expired. These replicas are taken by the reaper and they are deleted. 

> Sometimes, replicas can become unhealthy. A dump is created by the dumper daemon. The auditor checks these dumps and declares replicas 

Reviewer: ```suggestion
Sometimes, replicas can become unhealthy. A dump is created by the dumper daemon. The auditor checks these dumps and declares replicas.
```

## How is a replica declared bad?

> - An operator can declare a replica as bad issuing rucio cli commands. These bad replicas are taken by the necromancer daemon and then deleted if they cannot be recovered from other RSEs.

Reviewer: Can you add an example code block? It's useful for operators so they don't have to go diving into document.

> - the suspicios-replica-recoverer is a daemon that will analyze different counters (transfer errors, download errors, etc ...) to mark replicas as being suspicious to have an issue. After hitting

Reviewer: ```suggestion
- The suspicious-replica-recoverer is a daemon that will analyze different counters (transfer errors, download errors, etc ...) to mark replicas as being suspicious to have an issue. After hitting
```

Reviewer: Still mispelled

a certain limit (configurable), the replica is marked as bad and eventually consumed by the necromancer.

## What is the purpose of undertaker?

> A dataset is "never" deleted, however, when the dataset is known to be spurious, there is no point having it in the catalog. The undertaker daemon takes care to remove these datasets.

Reviewer: "spurious" is an odd worse choice, I'd just call it "bad" like the status defines it.

> A dataset is "never" deleted, however, when the dataset is known to be bad, there is no point having it in the catalog. The undertaker daemon takes care to remove these datasets.

Reviewer: ```suggestion
A dataset is never "deleted", however, when the dataset is known to be bad, there is no point having it in the catalog. The undertaker daemon takes care to remove these datasets.
```
"deleted" is the specific term that isn't actually done

---
id: daemons
title: Rucio Daemons
sidebar_label: daemons
---

Rucio relies on several daemons (processes) to perform different logic.
Most of the daemons connect to the DB to read some data, perform some computation, 
and then write some data back into the DB.

Usually one daemon will create some work for another daemon and viceversa.
In Rucio realm, daemons communicate to others by the DB.

> The following table reprensents a high level view of the responsibility of each of the daemons.

Reviewer 2: ```suggestion
The following table represents a high level view of the responsibility of each of the daemons.
```

## Daemons

Name | Domain|  Purpose |  Details
-----------| ----------------- | -------| ----- |

> rucio-abacus-account | Accounting | Account usage | [Details](https://rucio.cern.ch/documentation/bin/rucio-abacus-account)

Reviewer: ```suggestion
rucio-abacus-account | Accounting | Updates account usage counters | [Details](https://rucio.cern.ch/documentation/bin/rucio-abacus-account)
```

rucio-abacus-collection-replica | Accounting |  Updates collection replicas | [Details](https://rucio.cern.ch/documentation/bin/rucio-abacus-collection-replica)
rucio-abacus-rse | Accounting | Updates RSE counters | [Details](https://rucio.cern.ch/documentation/bin/rucio-abacus-rse)

> rucio-atropos | Replica | End the life of the rules according to the Lifetime Model | [Details](https://rucio.cern.ch/documentation/bin/rucio-atropos)

Reviewer: ```suggestion
rucio-atropos | Replica | Deletes rules that have reached end-of-life according to the Lifetime Model | [Details](https://rucio.cern.ch/documentation/bin/rucio-atropos)
```

> rucio-auditor | Replica | Find inconsistencies on storage, for example, dark data discovery | [Details](https://rucio.cern.ch/documentation/bin/rucio-auditor)

Reviewer: ```suggestion
rucio-auditor | Replica | Finds inconsistencies on storage, for example, dark data discovery | [Details](https://rucio.cern.ch/documentation/bin/rucio-auditor)
```

rucio-automatix | Replica | Used for testing: injects random data in RSEs to check liveness | [Details](https://rucio.cern.ch/documentation/bin/rucio-automatix)
rucio-bb8 | Replica | Rebalance data across RSEs | [Details](https://rucio.cern.ch/documentation/bin/rucio-bb8)

> rucio-c3po | Replica | Places data according to dynamic rules: available space, popularity, etc ... | [Details](https://rucio.cern.ch/documentation/bin/rucio-c3po)

Reviewer 2: AFAIK C3PO should be removed as it is not used / supported anymore: https://github.com/rucio/rucio/issues/6872#issuecomment-2220853416

Should we keep it in the docs for now?

rucio-cache-client | Replica | Populates information of replicas on volatile storage  | [Details](https://rucio.cern.ch/documentation/bin/rucio-cache-client)
rucio-cache-consumer | Replica | Adds and deletes cache replicas to the Rucio catalog | [Details](https://rucio.cern.ch/documentation/bin/rucio-cache-consumer)
rucio-conveyor-finisher | Transfer | Updates Rucio internal state after the file transfer has finished | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-finisher)
rucio-conveyor-poller | Transfer | Polls updates from the transfer tool to check the transfer state | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-poller)
rucio-conveyor-preparer | Transer | Prepares data transfers | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-preparer)

> rucio-conveyor-receiver | Transfer | Sister of poller, instead of polling for updates, it reads transfer tools notifications to check transfer state | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-receiver)

Reviewer: ```suggestion
rucio-conveyor-receiver | Transfer | Reads transfer tools notifications to check transfer state. Sister of conveyor-poller | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-receiver)
```

> rucio-conveyor-receiver | Transfer | Sister of polller, instead of polling for updates, it reads transfer tools notifications to check transfer state | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-receiver)

Reviewer 2: ```suggestion
rucio-conveyor-receiver | Transfer | Sister of poller, instead of polling for updates, it reads transfer tools notifications to check transfer state | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-receiver)
```

Could be nicer to say "similar to" or "companion to" instead of "sister of"

> rucio-conveyor-stager | Transfer| Issues staging (bring online) requests to tape RSEs  | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-stager)

Reviewer 2: ```suggestion
rucio-conveyor-stager | Transfer | Issues staging (bring online) requests to tape RSEs  | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-stager)
```

> rucio-conveyor-submitter | Transfer | Submit transfer requests to the transfer tool (prepares the transfer as well if the conveyor-preparer is not enabled) | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-submitter)

Reviewer: ```suggestion
rucio-conveyor-submitter | Transfer | Submits transfer requests to the transfer tool (prepares the transfer as well if the conveyor-preparer is not enabled) | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-submitter)
```

> rucio-conveyor-submitter | Transfer| Submit transfer requests to the transfer tool (prepares the transfer as well if the converyor-preparer is not enabled) | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-submitter)

Reviewer 2: ```suggestion
rucio-conveyor-submitter | Transfer | Submit transfer requests to the transfer tool (prepares the transfer as well if the conveyor-preparer is not enabled) | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-submitter)
```

> rucio-conveyor-throttler | Transfer | Queues transfer requests inside Rucio, applying limits, ex: ony one transfer at a time, etc ... | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-throttler)

Reviewer 2: ```suggestion
rucio-conveyor-throttler | Transfer | Queues transfer requests inside Rucio, applying limits, ex: only one transfer at a time, etc ... | [Details](https://rucio.cern.ch/documentation/bin/rucio-conveyor-throttler)
```

> rucio-dark-reaper | Deletion | Deletes quarantained replicas | [Details](https://rucio.cern.ch/documentation/bin/rucio-dark-reaper)

Reviewer 2: ```suggestion
rucio-dark-reaper | Deletion | Deletes quarantined replicas | [Details](https://rucio.cern.ch/documentation/bin/rucio-dark-reaper)
```

rucio-dumper | Consistency | Dumps file lists. The rucio-auditor consumes these dumps to discover dark data | [Details](https://rucio.cern.ch/documentation/bin/rucio-dumper)
rucio-follower | Telemetry | Aggregates events affecting DIDs | [Details](https://rucio.cern.ch/documentation/bin/rucio-follower)

> rucio-hermes | Telemtry | Sends Rucio messages to external services (influxDB, OpenSearch, ActiveMQ, ...) | [Details](https://rucio.cern.ch/documentation/bin/rucio-hermes)

Reviewer 2: ```suggestion
rucio-hermes | Telemetry | Sends Rucio messages to external services (InfluxDB, OpenSearch, ActiveMQ, ...) | [Details](https://rucio.cern.ch/documentation/bin/rucio-hermes)
```

rucio-judge-cleaner | Rule | Cleans expired replication rules | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-cleaner)
rucio-judge-evaluator | Rule | Creates and evaluates replication rules based on their state (OK/REPL/STUCK)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-evaluator)
rucio-judge-injector | Rule | Asynchronously injects replication rules  | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-injector)
rucio-judge-repairer | Rule | Repairs stuck replication rules (STATE=STUCK) | [Details](https://rucio.cern.ch/documentation/bin/rucio-judge-repairer)

> rucio-kronos | Telemetry | Consumes Rucio tracing messages and updates access time of replicas | [Details](https://rucio.cern.ch/documentation/bin/rucio-kronos)

Reviewer 2: Also updates the access count on the DID, not sure if we want to mention that though

rucio-minos | Replica | Reads list of physical file names (PFNs) declared bad and classifies them in: temporary unavailable  and permanently unavailable (to be recovered by the necromancer daemon)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-minos)

> rucio-minos-temporary-expiration | Replica | Moves back replicas TEMPORARY_UNAVAILABLE into AVAILABE state  | [Details](https://rucio.cern.ch/documentation/bin/rucio-minos-temporary-expiration)

Reviewer 2: ```suggestion
rucio-minos-temporary-expiration | Replica | Moves back replicas TEMPORARY_UNAVAILABLE into AVAILABLE state  | [Details](https://rucio.cern.ch/documentation/bin/rucio-minos-temporary-expiration)
```

> rucio-necromancer | Deletion | Works on permantenly unavailable replicas, it tries to recover the data from other valid replicas if any, else declares the replica as lost  | [Details](https://rucio.cern.ch/documentation/bin/rucio-necromancer)

Reviewer 2: ```suggestion
rucio-necromancer | Deletion | Works on permanently unavailable replicas, it tries to recover the data from other valid replicas if any, else declares the replica as lost  | [Details](https://rucio.cern.ch/documentation/bin/rucio-necromancer)
```

> rucio-oauth-manager | Auth/Authz | deletes expired access token and and oath sessions and refresh access tokens with refresh tokens  | [Details](https://rucio.cern.ch/documentation/bin/rucio-oauth-manager)

Reviewer 2: ```suggestion
rucio-oauth-manager | Auth/Authz | deletes expired access token and and OAuth sessions and refresh access tokens with refresh tokens  | [Details](https://rucio.cern.ch/documentation/bin/rucio-oauth-manager)
```

> rucio-oauth-manager | Auth/Authz | deletes expired access token and and oath sessions and refresh access tokens with refresh tokens  | [Details](https://rucio.cern.ch/documentation/bin/rucio-oauth-manager)

Reviewer: ```suggestion
rucio-oauth-manager | Auth/Authz | Deletes expired access token and and OAuth sessions and refresh access tokens with refresh tokens  | [Details](https://rucio.cern.ch/documentation/bin/rucio-oauth-manager)
```

rucio-reaper | Deletion | Deletes replicas that dont' have locks anymore, i.e they have a tombstone set | [Details](https://rucio.cern.ch/documentation/bin/rucio-reaper)

> rucio-suspicious-replica-recoverer | Replica | Declares suspicios replicas as bad if they are found available on other RSEs, so necromancer will work on them | [Details](https://rucio.cern.ch/documentation/bin/rucio-replica-recoverer)

Reviewer 2: ```suggestion
rucio-suspicious-replica-recoverer | Replica | Declares suspicious replicas as bad if they are found available on other RSEs, so necromancer will work on them | [Details](https://rucio.cern.ch/documentation/bin/rucio-replica-recoverer)
```

> rucio-rse-decommissioner | Deletion | decomissions an RSE. The actions to perform are specified in decomissioning profiles (delete all data, move replicas, etc ...)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-rse-decomissioner)

Reviewer: ```suggestion
rucio-rse-decommissioner | Deletion | Decomission an RSE. The actions to perform are specified in decomissioning profiles (delete all data, move replicas, etc ...)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-rse-decomissioner)
```

Reviewer: Link to info on decomissioning profile docs would also be helpful

rucio-storage-consistency-actions | Consistency | Applies corrective actions as a result of a consistency check on an RSE | [Details](https://rucio.cern.ch/documentation/bin/rucio-storage-consistency)
rucio-transmogrifier | Rule | Creates replication rules for DIDs matching a subscription | [Details](https://rucio.cern.ch/documentation/bin/rucio-transmogrifier)
rucio-undertaker | Deletion | Manages expired DIDs, deleting them (does not delete replicas)  | [Details](https://rucio.cern.ch/documentation/bin/rucio-undertaker)

## FAQ

### Conveyor daemons
It is important to know the following. The throttler daemon will need the preparer to work.

> The preparer is a daemon that optimized transfer requests, while recommended to install, it's not mandatory.

Reviewer 2: ```suggestion
The preparer is a daemon that optimizes transfer requests, while recommended to install, it's not mandatory.
```

The submitter is the only daemon needed to submit transfers and can do a subset of what the preparer can do.
To update the state of requests, the conveyor poller (polls for changes) or conveyor receiver (listens for changes)  are needed to understand the new state.

> The finisher analyzes this new state and updates the state.

Reviewer: Going a step further, this would be easier to read if the sentences were shorter

> ### What happens when a ruile is stuck?

Reviewer 2: ```suggestion
### What happens when a rule is stuck?
```

The judge repairer will analyze why the transfer is stuck and try to unstuck it, eventually resubmitting the request.

> ### What happens when new data is added to an existing dataset that was already copied (replicas created)?

Reviewer: ```suggestion
### What happens when new data is added to an existing dataset that already has replicas?
The existing dataset already has copies on different RSEs. 
```

The judge evaluator will keep track of new data added to datasets that are already replicated to trigger the necessary transfer requests to ensure all new data is copied to the RSEs.

> ### What is the purpose of the minion daemons?

Reviewer 2: ```suggestion
### What is the purpose of the minos daemons?
```

> An human operator can declare some datasets temporary unavailable due to maintenance, outages, etc ... The operator will set an expiration time. 

Reviewer 2: ```suggestion
An human operator can declare some datasets as temporarily unavailable due to maintenance, outages, etc ... The operator will set an expiration time. 
```

> An human operator can declare some datasets temporary unavailable due to maintenance, outages, etc ... The operator will set an expiration time. 

Reviewer: ```suggestion
An human operator can declare some datasets as temporarily unavailable due to maintenance, outages, etc ... The operator will set an expiration time on the temporary unavailable status. 
```

When the expiration time is reached, the minos-temporary-expiration will put the replicas back in available state.

### What is the relationship between auditor, rucio-dumper and dark-reaper?

> The dumper will create a dump of all the files in an RSE that will be passed to the auditor. The auditor will check for inconsistencies and mark missing data as dark data (quarantained replicas).

Reviewer 2: ```suggestion
The dumper will create a dump of all the files in an RSE that will be passed to the auditor. The auditor will check for inconsistencies and mark missing data as dark data (quarantined replicas).
```

> Dark reaper is the one deleting this dark data to free up space from the quarantained replicas table.

Reviewer 2: ```suggestion
Dark reaper is the one deleting this dark data to free up space from the quarantined replicas table.
```

> ### How data is deleted?

Reviewer 2: ```suggestion
How is data deleted?
```

> ### How data is deleted?

Reviewer 2: ```suggestion
### How is data deleted?
```

When replicas are healthy, the judge-cleaner will set a tombstone on replicas where the lifetime has expired. These replicas are taken by the reaper and they are deleted. 
Sometimes, replicas can become unhealthy. A dump is created by the dumper daemon. The auditor checks these dumps and declares replicas 

> ## How to declare replicas as bad?

Reviewer: ```suggestion
## How is a replica declared bad?
```

### What happens when new data is added to an existing dataset that already has replicas?
The judge evaluator will keep track of new data added to datasets that are already replicated to trigger the necessary transfer requests to ensure all new data is copied to the RSEs.

### What is the purpose of the minos daemons?
An human operator can declare some datasets as temporarily unavailable due to maintenance, outages, etc ...The operator will set an expiration time on the temporary unavailable status. When the expiration time is reached, the minos-temporary-expiration will put the replicas back in available state.

### What is the relationship between auditor, rucio-dumper and dark-reaper?
The dumper will create a dump of all the files in an RSE that will be passed to the auditor. The auditor will check for inconsistencies and mark missing data as dark data (quarantined replicas).
Dark reaper is the one deleting this dark data to free up space from the quarantined replicas table.

### How is data deleted?
When replicas are healthy, the judge-cleaner will set a tombstone on replicas where the lifetime has expired. These replicas are taken by the reaper and they are deleted. 

> Sometimes, replicas can become unhealthy. A dump is created by the dumper daemon. The auditor checks these dumps and declares replicas 

Reviewer: I think you lost the end of a sentence here. Declares replicas as what?

## How is a replica declared bad?
- An operator can declare a replica as bad issuing rucio cli commands.
```
rucio-admin replicas declare-bad [-h] --reason REASON [--inputfile [INPUTFILE]] [--allow-collection] [--lfns [LFNS]] [--scope [SCOPE]] [--rse [RSE]] [listbadfiles ...]
```

These bad replicas are taken by the necromancer daemon and then deleted if they cannot be recovered from other RSEs.
- the suspicious-replica-recoverer is a daemon that will analyze different counters (transfer errors, download errors, etc ...) to mark replicas as being suspicious to have an issue. After hitting

> a certain limit (configurable), the replica is marked as bad and eventually consumed by the necromancer.

Reviewer: Configurable how? This should call out the exact parameter used
